core/src/main/java/org/apache/iceberg/variants/VariantUtil.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.variants;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+import java.util.function.Function;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+class VariantUtil {
+  private static final int BASIC_TYPE_MASK = 0b11;
+  private static final int BASIC_TYPE_PRIMITIVE = 0;
+  private static final int BASIC_TYPE_SHORT_STRING = 1;
+  private static final int BASIC_TYPE_OBJECT = 2;
+  private static final int BASIC_TYPE_ARRAY = 3;
+
+  private VariantUtil() {}
+
+  /** A hacky absolute put for ByteBuffer */
+  static int writeBufferAbsolute(ByteBuffer buffer, int offset, ByteBuffer toCopy) {
+    int originalPosition = buffer.position();
+    buffer.position(offset);
+    ByteBuffer copy = toCopy.duplicate();
```

Comment:
This feels less than ideal.  Are we trying to avoid `mark/position/reset`?  Are we expecting implementations wouldn't support that?

Reply:
I think this should avoid modifying any state in the buffer because it could be passed in. While `mark` does a similar thing and saves the position, it still modifies the state of the buffer because there can be only one mark in use at a time.

We also use this `ByteBuffer#duplicate` pattern in lots of places where we need to pass a buffer into a method that will modify `position`. We've verified that it is a cheap operation and can be used without a performance penalty, even in tight loops on the read path.